Fix answer saving for quizzes, practice quizzes, and surveys

rtibbles opened the pull request:
## Summary
* Removes the debounce save behaviour that has caused issues with multiple attempts being recorded
* Updates to only saving when changing question or submitting the quiz
* Does an optimistic update of current question answer state with an additional tracking ref
* Does a flyby fix of polling on the class assignment page, switches to useTimeout and passes in the classId value instead of the computed object. 

## References
Fixes https://github.com/learningequality/kolibri/issues/9444
Fixes issue noted in https://github.com/learningequality/kolibri/issues/13599

## Reviewer guidance
Engage with:

A coach created quiz
A practice quiz
A survey

Ensure in each case that responses are correctly recorded, and that the UI updates appropriately (showing total number of questions answered, correctly filled in blob indicators etc). Check reports for data persistence.

Go to a learner's class page, ensure that the backend is polled roughly every 30 seconds and it doesn't error.

rtibbles committed:
Defer answer saving until after a user has navigated to another question or submitted the quiz.

rtibbles committed:
Fix polling on the class assignments page.

marcellamaki reviewed (commented):
I've done a first code read through and initial manual QA confirmation. I was able to reproduce the survey submission issue on the current 0.18.x and confirm the fix that @rtibbles reports, but @pcenov could you please double check (with the more thorough QA approach 🙂) and confirm the fix works for you, and that you don't encounter any other problems? We should also regression test submission for quizzes and practice quizzes -- just to make sure all is still working as expected. 

I'll also do a final code review in the morning

(Note this doesn't address the issue Radina mentions of the timeframe of the attempts not being in sync with each other in the quiz reports)

ETA: i have reread the code this morning and do not see any readability/approach issues, so will wait for QA approval of the fix

pcenov commented:
Hi @rtibbles and @marcellamaki, observed the following issues:

1. All unanswered questions in a quiz/practice quiz/survey are counted as answered:

https://github.com/user-attachments/assets/30932bf3-91d2-42c4-b3f8-be75d4ddaa69

2. Sometimes the answer to a survey is not saved at all - this happens when I go to the next question and then go back to the previous one and then back to the next question:

https://github.com/user-attachments/assets/f159d3c7-cfd5-497a-95c2-1ff5f0c601dc

3. Navigation between the questions in a practice quiz is too slow now to the extent that the user feels that nothing has happened after having clicked on a question:

https://github.com/user-attachments/assets/7f5f5b0d-ffc8-41b6-8793-758325b5d6e4

4. After submitting a quiz I noticed that some of the correct answers are counted as incorrect. This is also caused by having navigated back and forth before giving the correct answer:

https://github.com/user-attachments/assets/b2db165b-59b1-4578-b1dd-2572c4c3c9b8

5. Similarly seeing empty attempts:

https://github.com/user-attachments/assets/96349544-5212-494b-ba8b-87e5a08edffa

[kolibri-home-folder.zip](https://github.com/user-attachments/files/21511404/kolibri-home-folder.zip)

marcellamaki commented:
Thanks @pcenov this is very helpful -- we'll think through what could be going on

rtibbles commented:
Just to clarify number 3:

> Navigation between the questions in a practice quiz is too slow now to the extent that the user feels that nothing has happened after having clicked on a question:

From the screen recording this appears to be a coach created quiz, not a practice quiz? (practice quizzes are specifically those that users interact with directly in a channel without being assigned, as in the QA channel)

pcenov commented:
Yes, sorry - it's a coach created quiz.

marcellamaki commented:
I've taken another look at this and did a quick manual QA which seems to confirm the changes that Richard made here. However, since @pcenov previously found that my "this fixes is!" was not true 😄 I will again defer to him to confirm the fix. But it should be ready for re-review, QA team. Thank you

pcenov commented:
Hi @rtibbles and @marcellamaki, I confirm that the points listed above are fixed now and I was able to identify just one more issue while regression testing - it's not possible to edit any of the given answers in a survey. This is valid for all question types not just the free response:

https://github.com/user-attachments/assets/aae46d10-9bb4-4257-baf0-238676fb95d2

marcellamaki commented:
hmmm, tricky. well that is not quite there then. Thank you @pcenov

rtibbles committed:
Fix issue with practice quiz and survey randomization caused by misreference of userId

rtibbles committed:
Display attempts in reverse chronological order for non-exercises so that we default to the most recent.

rtibbles commented:
OK, I think I have fixed the issue. It appeared to be a problem with the ordering of attempts in reporting.

As such, I have changed how we order attempts for individual questions in reports for quizzes (practice and coach created) and surveys. Now the "final attempt" which is the "given" attempt is displayed on initial page load (and the attempts overall are shown in reverse chronological order).

By contrast the exercise reports still show the attempts in chronological order, with the first given answer being selected by default, as that is the "given" answer for that question.

rtibbles commented:
This will require review of the learner and coach reports for coach quizzes, practice quizzes, and surveys, and of the coach reports for exercises in lessons.

marcellamaki commented:
> I've taken another look at this and did a quick manual QA which seems to confirm the changes that Richard made here. However, since @pcenov previously found that my "this fixes is!" was not true 😄 I will again defer to him to confirm the fix. But it should be ready for re-review, QA team. Thank you

just this. again. :) i only looked at the most obvious example of the case described here and haven't done any other regression testing.

pcenov reviewed (approved):
Thanks @marcellamaki and @rtibbles - I confirm that this is implemented as specified in Richard's latest comment, no new issues observed while regression testing.
